src/java.base/unix/native/libnet/PlainDatagramSocketImpl.c

```diff
@@ -1996,6 +1996,25 @@ static void mcast_join_leave(JNIEnv *env, jobject this,
         if (setsockopt(fd, IPPROTO_IP, (join ? IP_ADD_MEMBERSHIP:IP_DROP_MEMBERSHIP),
                        (char *) &mname, mname_len) < 0) {
 
+#ifdef __APPLE__
+            if (errno == ENOMEM) {
+                if (setsockopt(fd, IPPROTO_IP, (join ? IP_ADD_MEMBERSHIP:IP_DROP_MEMBERSHIP),
+                           (char *) &mname, mname_len) < 0 ) {
+                        if (errno) {
+                            if (join) {
+                                NET_ThrowCurrent(env, "setsockopt IP_ADD_MEMBERSHIP failed");
+                            } else {
+                                if (errno == ENOENT)
+                                    JNU_ThrowByName(env, JNU_JAVANETPKG "SocketException",
+                                        "Not a member of the multicast group");
+                                else
+                                    NET_ThrowCurrent(env, "setsockopt IP_DROP_MEMBERSHIP failed");
+                            }
+                            return;
+                        }
+                }
+            } else {
+#endif
```

Comment:
The handling of ENOMEM here is consistent with how it is handled in the NIO area - good.

I wonder if a little restructuring may simplify the call flow? For example, something similar to:
```
  int n;
  ...
  /*
   * Join the multicast group.
   */
   n = setsockopt(fd, IPPROTO_IP, (join ? IP_ADD_MEMBERSHIP:IP_DROP_MEMBERSHIP),
                          (char *) &mname, mname_len);
#ifdef __APPLE__
   // workaround macOS bug where IP_ADD/DROP_MEMBERSHIP fails intermittently
   if (n < 0 && errno == ENOMEM) {
      n = setsockopt(fd, IPPROTO_IP, (join ? IP_ADD_MEMBERSHIP:IP_DROP_MEMBERSHIP),
                                          (char *) &mname, mname_len);
   }
#endif

  if (n < 0) {  ...
```

Reply:
yes, I'll do that ... originally I did this but reverted to current change to retain the current structure of the file
thanks for the suggestion